apps/api-extractor/src/generators/PackageTypingsGenerator.ts

```diff
@@ -0,0 +1,402 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+/* tslint:disable:no-bitwise */
+
+import * as fs from 'fs';
+import * as ts from 'typescript';
+
+import { ExtractorContext } from '../ExtractorContext';
+import IndentedWriter from '../IndentedWriter';
+import TypeScriptHelpers from '../TypeScriptHelpers';
+import { Span } from './Span';
+
+/**
+ * An "Entry" is a type definition that we encounter while traversing the
+ * references from the package entry point.  This data structure helps filter,
+ * sort, and rename the entries that end up in the output package typings file.
+ */
+class Entry {
+  /**
+   * The original name of the symbol, as exported from the module (i.e. source file)
+   * containing the original TypeScript definition.
+   */
+  public localName: string;
+
+  /**
+   * The localName, possibly renamed to ensure that all the top-level exports have unique names.
+   */
+  public uniqueName: string | undefined = undefined;
+
+  /**
+   * The compiler symbol where this type was defined, after following any aliases.
+   */
+  public followedSymbol: ts.Symbol;
+
+  /**
+   * If true, this entry should be emitted using the "export" keyword instead of the "declare" keyword.
+   */
+  public exported: boolean = false;
+
+  private _sortKey: string|undefined = undefined;
+
+  public getSortKey(): string {
+    if (!this._sortKey) {
+      if (this.localName.substr(0, 1) === '_') {
+        // Removes the leading underscore, for example:  "_example" --> "example*"
+        // This causes internal definitions to sort alphabetically with regular definitions.
+        // The star is appended to preserve uniqueness, since "*" is not a legal  identifier character.
+        this._sortKey = this.localName.substr(1) + ' ';
+      } else {
+        this._sortKey = this.localName;
+      }
+    }
+    return this._sortKey;
+  }
+}
+
+/**
+ * Return value for PackageTypingsGenerator._followAliases()
+ */
+interface IFollowAliasesResult {
+  /**
+   * The original symbol that defined this entry, after following any aliases.
+   */
+  symbol: ts.Symbol;
+
+  /**
+   * If true, the symbol was declared by an external package, e.g. versus being imported
+   * from another source file in the current project.
+   */
+  external: boolean;
+
+  /**
+   * If true, the symbol is exported from a module somewhere.  If false, then it's
+   * a global ambient definition or else a private declaration in the referencing file.
+   * NOTE: The "external" status is unknown if moduleExport=true.
+   */
+  moduleExport: boolean;
+}
+
+export default class PackageTypingsGenerator {
```

Comment:
>PackageTypingsGenerator [](start = 21, length = 23)

Where is the test for this class?

Reply:
I need to figure out how to reenable the API Extractor tests using Jest.  I'll do that in a separate PR.

---
In reply to: [152429942](https://github.com/Microsoft/web-build-tools/pull/448#discussion_r152429942) [](ancestors = 152429942)

Reply:
I've started moving the integration test projects into the "/build-tests" folder.

---
In reply to: [152429942](https://github.com/Microsoft/web-build-tools/pull/448#discussion_r152429942) [](ancestors = 152429942)